app-ios/tutanota/Sources/Sharing/SharingStorage.swift

```diff
@@ -0,0 +1,176 @@
+import Foundation
+
+let TUTANOTA_APP_GROUP = "group.de.tutao.tutanota"
+let TUTANOTA_SHARE_SCHEME = "tutashare://"
+
+/// this gets shared to the main app and contains all the info
+/// to create a new mail & the cleanup
+struct SharingInfo: Codable {
+  /// this is the key that the information is stored under in the
+  /// UserDefaults and on disk
+  var timestamp: String
+  /// body text of new mail
+  var text: String
+  /// files to attach
+  var fileUrls: [URL]
+}
+
+/// different errors that can happen when trying to load, convert and write
+/// shared items to the app group storage
+enum SharingError: Error {
+  case failedToLoad
+  case failedToWrite
+  case failedToRead
+}
+
+/// different types of items are shared under different names
+/// ident identifies the type, content the associated data
+enum SharedItem {
+  case fileUrl(ident: String, content: URL?)
+  case image(ident: String, content: UIImage?)
+  case text(ident: String, content: String?)
+  case contact(ident: String, content: String?)
+  
+  func ident() -> String {
+    switch self {
+    case .fileUrl(ident: let ident, content: _), .text(ident: let ident, content: _), .contact(ident: let ident, content: _), .image(ident: let ident, content: _):
+      return ident
+    }
+  }
+}
+
+/// there is a bunch of uniform type identifiers that can be set by the sharing app. it's not clear what's available,
+/// but there are some constants defined available with "import UniformTypeIdentifiers".
+func stringToItemType(ident: String) -> SharedItem? {
+  switch ident {
+  case
+    "public.png", "public.jpeg", "public.tiff", //shared from photos
+    "public.file-url" // shared from files
+    :
+    return .fileUrl(ident: ident, content: nil)
+  case
+    "public.image" // shared from e.g. Signal, no URL but image data
+    :
+    return .image(ident: ident, content: nil)
+  case
+    "public.url", // shared image from safari, shared link, shared pdf...
+    "public.plain-text" // shared from Notes, safari
+    :
+    return .text(ident: ident, content: nil)
+  case
+    "public.vcard" // shared from contacts
+    :
+    return .contact(ident: ident, content: nil)
+  default:
+    return nil
+  }
+}
+
+/// write a single file to a subdirectory in the shared storage of the app group
+func writeToSharedStorage(subdir: String, name: String, content: Data) async throws -> URL {
+  return try await withCheckedThrowingContinuation { cont in
```

Comment:
why do we wrap it again here? We should either schedule it on bg thread with `Task` from the outside or from inside but `withCheckedThrowingContinuation` is only meant for bridging to callbacks.

Reply:
good catch, that's a remnant from a previous version